Sources/Util/Heap.swift

```diff
@@ -19,232 +19,176 @@
     }
 
     /**
-     * Returns the index of the parent of the element at index i.
-     * The element at index 0 is the root of the tree and has no parent.
+     * `delete` deletes the given value from this Heap.
      */
-    @inline(__always) internal func parentIndex(ofIndex index: Int) -> Int {
-        return (index - 1) / 2
-    }
+    func delete(_ node: HeapNode<K, V>) {
+        guard let targetIndex = self.nodes.firstIndex(where: { $0.value == node.value }),
+              let lastNode = self.nodes.popLast()
+        else {
+            return
+        }
 
-    /**
-     * Returns the index of the left child of the element at index i.
-     * Note that this index can be greater than the heap size, in which case
-     * there is no left child.
-     */
-    @inline(__always) internal func leftChildIndex(ofIndex index: Int) -> Int {
-        return 2 * index + 1
-    }
+        if targetIndex < 0 || self.length() == 0 {
+            return
+        }
 
-    /**
-     * Returns the index of the right child of the element at index i.
-     * Note that this index can be greater than the heap size, in which case
-     * there is no right child.
-     */
-    @inline(__always) internal func rightChildIndex(ofIndex index: Int) -> Int {
-        return 2 * index + 2
-    }
+        self.nodes[targetIndex] = lastNode
 
-    /**
-     * Returns the maximum value in the heap (for a max-heap) or the minimum
-     * value (for a min-heap).
-     */
-    func peek() -> T? {
-        return self.nodes.first
+        self.heapify(parentIndex: self.getParentIndex(targetIndex), targetIndex: targetIndex)
     }
 
     /**
-     * Adds a new value to the heap. This reorders the heap so that the max-heap
-     * or min-heap property still holds. Performance: O(log n).
+     * `push` pushes the given node onto this Heap.
      */
-    mutating func insert(_ value: T) {
-        self.nodes.append(value)
-        self.shiftUp(self.nodes.count - 1)
+    func push(_ node: HeapNode<K, V>) {
+        self.nodes.append(node)
+        self.moveUp(self.nodes.count - 1)
     }
 
     /**
-     * Adds a sequence of values to the heap. This reorders the heap so that
-     * the max-heap or min-heap property still holds. Performance: O(log n).
+     * `pop` removes and returns the maximum element from this Heap.
      */
-    mutating func insert<S: Sequence>(_ sequence: S) where S.Iterator.Element == T {
-        for value in sequence {
-            self.insert(value)
+    @discardableResult
+    func pop() -> HeapNode<K, V>? {
+        let count = self.nodes.count
+        guard let head = self.nodes[safe: 0] else {
+            return nil
         }
-    }
 
-    /**
-     * Allows you to change an element. This reorders the heap so that
-     * the max-heap or min-heap property still holds.
-     */
-    mutating func replace(index: Int, value: T) {
-        guard index < self.nodes.count else { return }
+        if count <= 0 {
+            return nil
```

Comment:
Due to line 96, it's unreachable

Reply:
Changed checking `nodes.count` to `isEmpty`. Thank you.